csrc/multidevice/utils.cpp

```diff
@@ -276,36 +276,88 @@ int64_t numDeviceDims(const TensorView* tv) {
 }
 
 namespace {
+std::pair<Val*, bool> computeIndex(
+    IterDomain* id,
+    const std::vector<IterDomain*>& sources,
+    std::unordered_map<IterDomain*, std::pair<Val*, bool>>& id_to_index) {
+  if (id == nullptr) {
+    return {nullptr, false};
+  }
 
-std::vector<IterDomain*> getInputsInTargetDomain(
-    IterDomain* loop_id,
-    const std::vector<IterDomain*>& target_domain) {
-  const std::vector<Val*> inputs_as_vals = IterVisitor::getInputsTo(
-      {loop_id}, {target_domain.begin(), target_domain.end()});
-
-  std::vector<IterDomain*> inputs_as_iter_domains;
-  inputs_as_iter_domains.reserve(inputs_as_vals.size());
-  std::transform(
-      inputs_as_vals.begin(),
-      inputs_as_vals.end(),
-      std::back_inserter(inputs_as_iter_domains),
-      [](Val* val) { return val->as<IterDomain>(); });
-  return inputs_as_iter_domains;
-}
+  std::vector<Expr*> transforms =
+      StmtSort::getExprsBetween({sources.begin(), sources.end()}, {id});
+  for (Expr* transform : transforms) {
+    if (std::all_of(
+            transform->outputs().begin(),
+            transform->outputs().end(),
+            [&](Val* val) {
+              return id_to_index.count(val->as<IterDomain>()) > 0;
+            })) {
+      continue;
+    }
 
-bool overlaps(
-    const std::vector<IterDomain*>& a,
-    const std::unordered_set<IterDomain*>& b) {
-  return std::any_of(
-      a.begin(), a.end(), [&](IterDomain* id) { return b.count(id); });
+    if (auto* split = dynamic_cast<Split*>(transform)) {
+      auto* in = split->in()->as<IterDomain>();
+      auto* outer = split->outer()->as<IterDomain>();
+      auto* inner = split->inner()->as<IterDomain>();
+
+      const auto& in_info = id_to_index.at(in);
+      id_to_index[outer] = {
+          div(in_info.first, inner->extent()), in_info.second};
```

Comment:
Is this assumed to be divisible?

Reply:
No -- it's very likely indivisible because in_info.first is the "index" and inner->extent() is the size.

Reply:
I asked this because I was assuming the traversal was backward.